atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/HostnamesByIpSupplier.java

```diff
@@ -43,20 +47,48 @@ public final class HostnamesByIpSupplier implements Supplier<Map<String, String>
     private static final String HOSTNAME_COLUMN = "hostname";
     private static final String IP_COLUMN = "ip";
 
-    private final Supplier<PoolingContainer<CassandraClient>> randomGoodHostSupplier;
+    private final Supplier<List<PoolingContainer<CassandraClient>>> hosts;
+    private final Duration timeout;
 
-    public HostnamesByIpSupplier(Supplier<PoolingContainer<CassandraClient>> randomGoodHostSupplier) {
-        this.randomGoodHostSupplier = randomGoodHostSupplier;
+    public HostnamesByIpSupplier(Supplier<List<PoolingContainer<CassandraClient>>> hosts) {
+        this(hosts, Duration.ofSeconds(60));
+    }
+
+    @VisibleForTesting
+    public HostnamesByIpSupplier(Supplier<List<PoolingContainer<CassandraClient>>> hosts, Duration timeout) {
+        this.hosts = hosts;
+        this.timeout = timeout;
     }
 
     @Override
     public Map<String, String> get() {
-        try {
-            return randomGoodHostSupplier.get().runWithPooledResource(getHostnamesByIp());
-        } catch (Exception e) {
-            log.warn("Could not get hostnames by ip from Cassandra", e);
-            return ImmutableMap.of();
+        List<PoolingContainer<CassandraClient>> containers = hosts.get();
+        Stopwatch timer = Stopwatch.createStarted();
+        for (PoolingContainer<CassandraClient> container : containers) {
+            try {
+                return container.runWithPooledResource(getHostnamesByIp());
+            } catch (Exception | Error e) {
+                log.warn(
+                        "Could not get hostnames by IP from Cassandra",
+                        SafeArg.of("poolSize", containers.size()),
+                        SafeArg.of("elapsed", timer.elapsed()),
+                        e);
+            }
+
+            if (timer.elapsed().toNanos() >= timeout.toNanos()) {
```

Comment:
nit:
```suggestion
            if (timer.elapsed().compareTo(timeout) >= 0) {
```

Reply:
I'll let @jeremyk-91 be the tie breaker here -- my preference is to keep `timer.elapsed().toNanos() >= timeout.toNanos()` for readability purposes.

Reply:
No strong preference. I'd probably have written the compareTo() version though 😅